### PR TITLE
EVG-14182 Add package-level docs

### PR DESCRIPTION
**Pull request by bsamek**

### CODE REVIEW

File: agent/doc.go [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Package agent runs tasks on build hosts. It polls the API server for a new
+// task, gets the configuration for that task, and then runs it.
```
Comment (Kimchelly) [inline agent/doc.go:2]: I think the second sentence isn't that relevant to the agent overall, unless we wanted to describe the different components of the agent in greater detail.

Does this deserve more explanation?

File: apimodels/doc.go [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// Package apimodels provides models used across multiple packages.
```
Comment (Kimchelly) [inline apimodels/doc.go:1]: Can we mention that it's for the application server's REST API?

File: model/doc.go [hunk 3]
```diff
@@ -0,0 +1,2 @@
+// Package model defines Evergreen's data model.
```
Comment (Kimchelly) [inline model/doc.go:1]: I think this one actually deserves more explanation since it's fairly complicated and has lots of random, unrelated things in it since it's _all_ the Evergreen DB models. I think this package comment is not that useful unless it actually enumerates some of the things here. For example:

* Can we call out what the subpackages represent?
* There's a lot of unrelated random things in the models (top-level) package. There's stuff related to task lifecycle,   project YAML config parsing (e.g. the parser project), github integration, etc.

File: repotracker/doc.go [hunk 4]
```diff
@@ -0,0 +1,2 @@
+// Package repotracker tracks GitHub repositories.
```
Comment (Kimchelly) [inline repotracker/doc.go:1]: Can we say what a repotracker is?

File: operations/doc.go [hunk 5]
```diff
@@ -0,0 +1,2 @@
+// Package operations defines the entry points for the application and CLI.
```
Comment (Kimchelly) [inline operations/doc.go:1]: Maybe I don't know what "entry points" refers to here, but can rephrase it as something like "commands for the Evergreen CLI, which serves are the binary for both the Evergreen server and CLI". Is it worth mentioning that the application server and client are one and the same CLI?

File: rest/doc.go [hunk 6]
```diff
@@ -1,39 +1,40 @@
-/*
-	The REST API V2 has a series of central types that are useful to understand when
-	adding new endpoints or increasing its functionality.
+/*Package rest provides an API for Evergreen users.
```
Comment (Kimchelly) [inline rest/doc.go:1]: Newline after `/*`

File: rest/doc.go [hunk 7]
```diff
@@ -1,39 +1,40 @@
...
+for attaching data to a request and a RequestHandler to execute the request.
 
-	RouteManagers are structs that define all of the functionality of a particular
-	API route, including definitions of each of the methods it implements and the
-	path used to access it.
+RouteManager
+
+RouteManagers are structs that define all of the functionality of a particular
+API route, including definitions of each of the methods it implements and the
+path used to access it.
```
Comment (Kimchelly) [inline rest/doc.go:38]: Side note: this comment is the most practically useful doc comment.

File: service/doc.go [hunk 8]
```diff
@@ -0,0 +1,2 @@
+// Package service exposes core UI and API endpoints.
```
Comment (Kimchelly) [inline service/doc.go:1]: Can we say it's the legacy API?

Is it worth mentioning that most API endpoints live in `rest/route` nowadays?

**Comment by bsamek** [comment]
I expect test failures, since it doesn't bump the agent version, and it will cause several packages that we don't lint to be linted, but there will not be any tests uploaded, which will trip the missing tests detector.

File: rest/doc.go [hunk 9]
```diff
@@ -1,39 +1,40 @@
...
-	database and service layer. It has two main sets of implementations: one that
-	communicate with the database and one that mocks the same functionality. Development
-	does not require the creation of new Connector's, only addition to extant ones.
+Connector
 
-	RequestHandler
+Connector is a very large interface that defines interaction with the backing
+database and service layer. It has two main sets of implementations: one that
+communicate with the database and one that mocks the same functionality. Development
+does not require the creation of new Connector's, only addition to extant ones.
```
Comment (Kimchelly) [inline rest/doc.go:17]: new Connectors

File: rest/doc.go [hunk 10]
```diff
@@ -1,39 +1,40 @@
...
-	completing an API request. It contains an Authenticator to handle access control.
-	Authenticator is an interface type with many implementations that permit user's
-	access based on different sets of criteria. MethodHandlers also contain functions
-	for attaching data to a request and a RequestHandler to execute the request.
+MethodHandler
 
-	RouteManager
+MethodHandler is an struct that contains all of the data necessary for
+completing an API request. It contains an Authenticator to handle access control.
+Authenticator is an interface type with many implementations that permit user's
```
Comment (Kimchelly) [inline rest/doc.go:30]: users'

File: rest/doc.go [hunk 11]
```diff
@@ -1,39 +1,40 @@
...
 
-	MethodHandler is an struct that contains all of the data necessary for
-	completing an API request. It contains an Authenticator to handle access control.
-	Authenticator is an interface type with many implementations that permit user's
-	access based on different sets of criteria. MethodHandlers also contain functions
-	for attaching data to a request and a RequestHandler to execute the request.
+MethodHandler
 
-	RouteManager
+MethodHandler is an struct that contains all of the data necessary for
```
Comment (Kimchelly) [inline rest/doc.go:28]: a struct

File: validator/doc.go [hunk 12]
```diff
@@ -0,0 +1,2 @@
+// Package validator checks that project configuration files are correct.
```
Comment (Kimchelly) [inline validator/doc.go:1]: It also validates distro settings.

File: agent/command/doc.go [hunk 13]
```diff
@@ -0,0 +1,2 @@
+// Package command defines the commands the agent can run.
```
Comment (Kimchelly) [inline agent/command/doc.go:1]: Can we clarify that the "commands" come from the project config?

File: thirdparty/doc.go [hunk 14]
```diff
@@ -0,0 +1,2 @@
+// Package thirdparty defines interactions with third-party services.
```
Comment (Kimchelly) [inline thirdparty/doc.go:1]: Can we mention that the cloud provider is not part of this?

As far as I've seen, this package is really vaguely defined and just contains a random mish-mash of unrelated glue code for stuff that's not written by us.

File: scheduler/doc.go [hunk 15]
```diff
@@ -0,0 +1,2 @@
+// Package scheduler orders tasks in distro queues.
```
Comment (Kimchelly) [inline scheduler/doc.go:1]: Is it worth mentioning the different scheduler implementations?

File: rest/doc.go [hunk 16]
```diff
@@ -1,39 +1,40 @@
...
 
-	Connector is a very large interface that defines interaction with the backing
-	database and service layer. It has two main sets of implementations: one that
-	communicate with the database and one that mocks the same functionality. Development
-	does not require the creation of new Connector's, only addition to extant ones.
+Connector
 
-	RequestHandler
+Connector is a very large interface that defines interaction with the backing
+database and service layer. It has two main sets of implementations: one that
```
Comment (Kimchelly) [inline rest/doc.go:15]: two main implementations

**Review by Kimchelly — CHANGES_REQUESTED** [review]

**Comment by Kimchelly** [comment]
If you only want to add a doc.go, you have to add them to the PRECIOUS package in the Makefile, or else they'll fail forever.

File: scheduler/doc.go (hunk 15 above)
Comment (bsamek) [reply scheduler/doc.go:1]: I don't think so, since we only support a single one.

File: service/doc.go (hunk 8 above)
Comment (bsamek) [reply service/doc.go:1]: I don't think that's true. Many of the API endpoints are still core parts of Evergreen, used by the agent, for example.

**Comment by bsamek** [comment]
> If you only want to add a doc.go to a package without any other files, you have to add them to the PRECIOUS package in the Makefile, or else they'll fail forever (i.e. mock and rest).

rest/doc.go already exists, so I don't think this is true.

**Review by Kimchelly — APPROVED** [review]
